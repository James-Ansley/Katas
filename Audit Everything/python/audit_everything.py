import abc
import re
from datetime import datetime
from pathlib import PurePath


# I have vaguely assumed this interface is for some reason largely unchangeable
# and is depended on beyond the scope of the audit manager
class FileSystem(abc.ABC):
    @abc.abstractmethod
    def get_files(self, dir_name: PurePath | str) -> list[str]:
        raise NotImplementedError

    @abc.abstractmethod
    def write_all_text(self, new_file: PurePath | str, content: str) -> None:
        raise NotImplementedError

    @abc.abstractmethod
    def read_all_lines(self, path: PurePath | str) -> list[str]:
        raise NotImplementedError

    def append(self, path: PurePath | str, content: str) -> None:
        self.write_all_text(
            path,
            "\n".join((
                *self.read_all_lines(path),
                content,
            ))
        )

    def num_lines_in_file(self, file) -> int:
        return len(self.read_all_lines(file))


class Visit:
    time_fmt = '%Y-%m-%d %H:%M:%S'

    def __init__(self, name: str, time: datetime):
        self.name = name
        self.time = time

    def format(self):
        return f"{self.name};{self.time.strftime(self.time_fmt)}"


class AuditManager:
    _audit_filename_template = "audit_{}.txt"
    _audit_filename_regex = r"audit_\d+.txt"

    def __init__(
          self,
          max_entries_per_file: int,
          directory_name: str,
          file_system: FileSystem,
    ):
        self._max_entries_per_file = max_entries_per_file
        self._directory = PurePath(directory_name)
        self._file_system = file_system

    def add_record(self, visitor_name: str, time_of_visit: datetime) -> None:
        self._file_system.append(
            self._next_available_file(),
            Visit(visitor_name, time_of_visit).format()
        )

    def _audit_files(self) -> list[PurePath]:
        paths = self._file_system.get_files(self._directory)
        paths = [PurePath(file) for file in paths]
        paths = [path for path in paths if path.is_relative_to(self._directory)]
        paths = [path for path in paths
                 if re.match(self._audit_filename_regex, path.name)]
        return sorted(paths)

    def _new_audit_file(self) -> PurePath:
        audit_file_number = len(self._audit_files()) + 1
        new_file = self._audit_filename_template.format(audit_file_number)
        return self._directory / new_file

    def _next_available_file(self) -> PurePath:
        file_paths = self._audit_files()
        has_capacity = (
              file_paths
              and (self._file_system.num_lines_in_file(file_paths[-1])
                   < self._max_entries_per_file)
        )
        if has_capacity:
            return file_paths[-1]
        else:
            return self._new_audit_file()
