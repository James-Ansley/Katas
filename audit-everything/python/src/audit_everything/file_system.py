import abc


class FileSystem(abc.ABC):
    @abc.abstractmethod
    def get_files(self, dir_name: str):
        raise NotImplementedError

    @abc.abstractmethod
    def write_all_text(self, new_file: str, content: str):
        raise NotImplementedError

    @abc.abstractmethod
    def read_all_lines(self, path: str):
        raise NotImplementedError
