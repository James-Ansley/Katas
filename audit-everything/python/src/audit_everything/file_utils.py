import os
import re
from pathlib import PurePath


class FileWrapper:
    def __init__(self, file_system, root):
        self.file_system = file_system
        self.root = root

    def list_matching(self, pattern: re.Pattern) -> list[str]:
        return [
            str(PurePath(path).relative_to(self.root))
            for path in self.file_system.get_files(self.root)
            if pattern.match(path)
        ]

    def file_length(self, path: str) -> int:
        return len(self.file_lines(path))

    def file_lines(self, path: str) -> list[str]:
        path = self.join_relative_to_root(path)
        if path in self.file_system.get_files(self.root):
            return self.file_system.read_all_lines(path)
        else:
            return []

    def append_to(self, path: str, line: str) -> None:
        self.file_system.write_all_text(
            self.join_relative_to_root(path),
            "\n".join((*self.file_lines(path), line)),
        )

    def join_relative_to_root(self, filename: str) -> str:
        return os.path.join(self.root, filename)
