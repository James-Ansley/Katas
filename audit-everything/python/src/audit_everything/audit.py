import re
from datetime import datetime
from typing import Final

from audit_everything.file_system import FileSystem
from audit_everything.file_utils import FileWrapper


class AuditManager:
    audit_regex: Final[re.Pattern] = re.compile(r".*audit_(\d+)\.txt")
    audit_format: Final[str] = "audit_{}.txt"
    record_fmt: Final[str] = "{};{:%Y-%m-%d %H:%M:%S}"

    def __init__(
          self,
          max_entries_per_file: int,
          directory_name: str,
          file_system: FileSystem,
    ):
        self._max_entries_per_file: Final[int] = max_entries_per_file
        self._files = FileWrapper(file_system, directory_name)

    def add_record(self, visitor_name: str, time_of_visit: datetime) -> None:
        record = self.record_fmt.format(visitor_name, time_of_visit)
        audit_path = self._next_available_path()
        self._files.append_to(audit_path, record)

    def _next_available_path(self) -> str:
        result = self._last_audit_path()
        if result is None:
            return self._format_audit_path(1)
        elif self._can_append_to(result):
            return result
        else:
            next_audit_number = self._audit_number_from_path(result) + 1
            return self._format_audit_path(next_audit_number)

    def _last_audit_path(self) -> str | None:
        files = self._files.list_matching(self.audit_regex)
        if files:
            return max(files, key=self._audit_number_from_path)
        else:
            return None

    def _can_append_to(self, audit_path) -> bool:
        return self._files.file_length(audit_path) < self._max_entries_per_file

    def _format_audit_path(self, number):
        return self.audit_format.format(number)

    @classmethod
    def _audit_number_from_path(cls, path):
        return int(cls.audit_regex.match(path).group(1))
