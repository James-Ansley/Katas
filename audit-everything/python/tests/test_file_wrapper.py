import re
from pprint import pformat

from approvaltests import verify

from audit_everything.file_utils import FileWrapper
from utils import InMemoryFileSystem, test


@test
def a_file_wrapper_returns_empty_defaults_for_non_existent_files():
    file_system = InMemoryFileSystem({})

    sut = FileWrapper(file_system, "root_dir")

    assert sut.list_matching(re.compile(r".*")) == []
    assert sut.file_length("a_file.txt") == 0
    assert sut.file_lines("a_file.txt") == []


@test
def a_file_wrapper_returns_the_content_of_files():
    file_system = InMemoryFileSystem({"root_dir/a_file.txt": "line 1\nline 2"})

    sut = FileWrapper(file_system, "root_dir")

    assert sut.list_matching(re.compile(r".*\.txt")) == ["a_file.txt"]
    assert sut.file_length("a_file.txt") == 2
    assert sut.file_lines("a_file.txt") == ["line 1", "line 2"]


@test
def a_file_wrapper_creates_a_new_file_when_appending_to_a_nonexistent_file():
    file_system = InMemoryFileSystem({})
    sut = FileWrapper(file_system, "root_dir")
    sut.append_to("a_file.txt", "some new text")
    verify(pformat(file_system.files))


@test
def a_file_wrapper_appends_to_existing_files():
    file_system = InMemoryFileSystem({"root_dir/a_file.txt": "line 1"})
    sut = FileWrapper(file_system, "root_dir")
    sut.append_to("a_file.txt", "some new text")
    verify(pformat(file_system.files))
