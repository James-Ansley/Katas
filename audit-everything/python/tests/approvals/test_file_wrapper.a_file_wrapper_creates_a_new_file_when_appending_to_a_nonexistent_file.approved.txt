{'root_dir/a_file.txt': 'some new text'}
